src/server/middleware/vulnerable/vulnerable.go

```diff
@@ -1,125 +1,138 @@
+		return scan.NewChecker()
+	}
+)
+
+// Middleware middleware which does the vulnerability prevention checking for the artifact in GET /v2/<name>/manifests/<reference> API
 func Middleware() func(http.Handler) http.Handler {
-	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-			doVulCheck, img, projectVulnerableSeverity, wl := validate(req)
-			if !doVulCheck {
-				next.ServeHTTP(rw, req)
-				return
-			}
-			rec := httptest.NewRecorder()
-			next.ServeHTTP(rec, req)
-			// only enable vul policy check the response 200
-			if rec.Result().StatusCode == http.StatusOK {
-				// Invalid project ID
-				if wl.ProjectID == 0 {
-					err := errors.Errorf("project verification error: project %s", img.ProjectName)
-					pkgE := internal_errors.New(err).WithCode(internal_errors.PROJECTPOLICYVIOLATION)
-					serror.SendError(rw, pkgE)
-					return
-				}
-
-				ctx := req.Context()
-				art, err := artifact.Ctl.GetByReference(ctx, img.Repository, img.Digest, nil)
-				if err != nil {
-					// TODO: error handle
-					return
-				}
-
-				cve := report.CVESet(wl.CVESet())
-				summaries, err := sc.DefaultController.GetSummary(ctx, art, []string{v1.MimeTypeNativeReport}, report.WithCVEWhitelist(&cve))
-
-				if err != nil {
-					err = errors.Wrap(err, "middleware: vulnerable handler")
-					pkgE := internal_errors.New(err).WithCode(internal_errors.PROJECTPOLICYVIOLATION)
-					serror.SendError(rw, pkgE)
-					return
-				}
-
-				rawSummary, ok := summaries[v1.MimeTypeNativeReport]
-				// No report yet?
-				if !ok {
-					err = errors.Errorf("no scan report existing for the artifact: %s:%s@%s", img.Repository, img.Tag, img.Digest)
-					pkgE := internal_errors.New(err).WithCode(internal_errors.PROJECTPOLICYVIOLATION)
-					serror.SendError(rw, pkgE)
-					return
-				}
-
-				summary := rawSummary.(*vuln.NativeReportSummary)
-
-				// Do judgement
-				if summary.Severity.Code() >= projectVulnerableSeverity.Code() {
-					err = errors.Errorf("current image with '%q vulnerable' cannot be pulled due to configured policy in 'Prevent images with vulnerability severity of %q from running.' "+
-						"Please contact your project administrator for help'", summary.Severity, projectVulnerableSeverity)
-					pkgE := internal_errors.New(err).WithCode(internal_errors.PROJECTPOLICYVIOLATION)
-					serror.SendError(rw, pkgE)
-					return
-				}
-
-				// Print scannerPull CVE list
-				if len(summary.CVEBypassed) > 0 {
-					for _, cve := range summary.CVEBypassed {
-						log.Infof("Vulnerable policy check: scannerPull CVE %s", cve)
-					}
-				}
+	return middleware.BeforeRequest(func(r *http.Request) error {
+		ctx := r.Context()
+
+		logger := log.G(ctx).WithFields(log.Fields{"middleware": "vulnerable"})
+
+		path := r.URL.EscapedPath()
+		art, err := artifactController.GetByReference(ctx, distribution.ParseName(path), distribution.ParseReference(path), nil)
```

Comment:
Can we use artifact Info as the single source of truth of a request?

Reply:
Changed to use artifact info